### ISSUE TITLE
Some questions about weights and padding

### ISSUE THREAD

**Issue by MVP-D77** [ShiqiYu/SimpleCNNbyCPP#1]
First, in `face_binary_cls.cpp`, such as `float conv0_weight[16*3*3*3]`,it means out_channels = 16, in_channels = 3, and size = 3.From teacher words, the weight data is RGB form.I want to know the data in the following 1-d vector, what is the form of data stored? Do it store  data one out_channels after another,?And for every out_channel, it store data R,G,B alternately or store all R data first and following G data and B data in the end?
`for example, one out_channel the data form is [r1,g1,b1,r2,g2,b2......,rn , gn ,bn] or [r1,r2.....,rn ,g1,g2.....,gn ,b1,b2.....,bn]?`

Second, in modle.py, for this part
```
self.backbone = nn.Sequential(
            ConvBNReLU(3, 16, 3, 2, 1),    # downsampled by 2, 128 -> 64
            nn.MaxPool2d(2, 2),            # downsampled by 2, 64 -> 32
            ConvBNReLU(16, 32, 3, 1),      # keep
            nn.MaxPool2d(2, 2),            # downsampled by 2, 32 -> 16
            ConvBNReLU(32, 32, 3, 2, 1)    # downsampled by 2, 16 -> 8
        )
```
for the forth line , it stride = 1 and padding is default value but in this situation, the out picture size will be two less than in size(30) with 3x3 conv_size .If we want to make the picture size is still 32, perhaps need 1 padding? And  from 128 to 64 with 1 padding and 2 stride, for calculate formula, out size is `(W-F+2P)/S+1` with W picture size = W, conv_size = F ,padding  = P, stride = S, it will be 64.5 with in picture we assume have a circle of 0s but we only use left and top ,the right and bottom will not be used as long as it becomes 64?

**Comment by fengyuentau** [comment]
1. The **input** is RGB format, and there is no RGB format for the weight.
2. The weight data in `face_binary_cls.cpp` is stored channel-wisely. Lets say the shape of a kernel is `[output_channel, input_channel, k, k]`, the address of the first sub-kernel is from `*kernel[0 * input_channel * k * k]` to `*kernel[1 * input_channel * k * k]`. And you should be able to get the result for the first `output_channel` based on the input and the first sub-kernel.
3. `ConvBNReLU(16, 32, 3, 1),      # keep` As you said, this layer does downsample to be 30, but this does not affect other parts. I will update the comments though. Please help me inform other students in the class since I am not in your group.
4. Not fully understand your last sentence. What does it mean by the 'circle of 0s', and the 'left and top, the right and bottom'? If  you are confused why it's 64 but not 64.5, [this page](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d) offers nice equations to compute the output shape given the input shape, kernel size, stride, and padding. The one you offered (`(W-F+2P)/S+1`) does not have the final round-to-bottom operation.

**Comment by fengyuentau** [comment]
Comments updated:
https://github.com/ShiqiYu/SimpleCNNbyCPP/blob/920ed7b3c194665ab1be12c7f4bfa0a30dd35f5e/model.py#L51-L57

**Comment by clover2024** [comment]
Hi @fengyuentau , I have a similar question about "how the weight data in face_binary_cls.cpp is stored" and "how the initial image data should be stored". 
1. I didn't fully understand what "channel-wisely" means. Is the weight data stored as 
`[i0o0k0, ... , i0o0k8, i0o1k0, ... , i0o1k8, .... ]`("channel-wisely" according to input channel) or `[i0o0k0, ... , i0o0k8, i1o0k0, ... , i1o0k8, .... ]`("channel-wisely" according to output channel) 
2. Should the 1-d initial input data be the form of `[r0c0R, r0c0G, r0c0B, r0c1R, r0c1G, r0c1B, ....]` or the form  of `[r0c0R, r0c1R, r0c2R, ... , r0c0G, r0c1G, r0c2G, ...., r0c0B, r0c1B, r0c2B, ...]` (channel-wisely)? (r0 means row0 and c0 means column0)
3. Similar to (2), Should the 1-d flatten output data (input to the fc layer) be the form of `[r0c0channel0, r0c0channel1, r0c0channel2, r0c0channel3,  ....]` or the form of `[r0c0channel0, r0c1channel0, r0c2channel0, r0c3channel0,  ....]`?
Thanks.

**Comment by fengyuentau** [comment]
@apostlewang 
1. I have already given a detailed example in readme, please check carefully. According to your words, the weight data is stored as `[o0i0k0, ... , o0i0k8, o0i1k0, ... , o0i1k8, .... ]` ("channel-wisely" according to output channel).
2. Please check OpenCV's official documentation first [imread](https://docs.opencv.org/4.5.1/d4/da8/group__imgcodecs.html#ga288b8b3da0892bd651fce07b3bbd3a56). It's stored as ***BGR*** format, and goes like `img[3*w*h]`. '3' means the 3-channel (BGR).
3. It depends on your code. We don't have any restrictions on your internal output data as far as I know, as long as your program works.

**Comment by clover2024** [comment]
@fengyuentau In the demo:

> Example of locating weights and bias for a 3x3 kernel at out_channels=o, in_channels=i:


```
for (int o = 0; o < out_channels; ++o) {
    for (int i = 0; i < in_channels; ++i) {
        // weights
        // first row of the kernel
        float kernel_oi_00 = conv0_weight[o*(in_channels*3*3) + i*(3*3) + 0];
        float kernel_oi_01 = conv0_weight[o*(in_channels*3*3) + i*(3*3) + 1];
        float kernel_oi_02 = conv0_weight[o*(in_channels*3*3) + i*(3*3) + 2];
        // and more rows ...

        // bias
        float bias_oi = conv0_bias[o];
    }
}
```
Should first row of the kernel be
```
 float kernel_oi_00 = conv0_weight[i*(out_channels*3*3) + o*(3*3) + 0];
 float kernel_oi_01 = conv0_weight[i*(out_channels*3*3) + o*(3*3) + 1];
 float kernel_oi_02 = conv0_weight[i*(out_channels*3*3) + o*(3*3) + 2];
```
if the weight data is stored as `[i0o0k0, ... , i0o0k8, i0o1k0, ... , i0o1k8, .... ]` ("channel-wisely" according to input channel)?

And for fc weights,
```
for (int o = 0; o < out_features; ++o) {
    for (int i = 0; i < in_features; ++i) {
        float w_oi = fc0_weight[o*out_features + i];
        // ...
    }
    float bias = fc0_bias[o];
}
```
if I didn't misunderstood, should the `float w_oi` be `float w_oi = fc0_weight[o*in_features + i];`?

**Comment by fengyuentau** [comment]
@apostlewang Good morning. I misunderstood your words, the dimension of `output_channels` should go first, then `input_channels`. I have corrected my previous comments.
